### PR TITLE
Setup test suite to stress bsp4s/bsp4j compatibility

### LINKED ISSUE #57: Write unit tests for bsp4s/bsp4j interrop
I noticed that there's minor typo differences between bsp4j and bsp4s

- `buildTarget/inverseSources` has wrong signature in bsp4j
- `ResourceItem.targets` should be `ResourceItem.target` in bsp4j

I'm writing a test suite to echo messages between bsp4j and bs4ps so we're sure typos are caught early.

### PR DESCRIPTION
**Pull request by olafurpg**
This commit adds a new test suite `TypoSuite` that catches
incompatibilies between bsp4s and bsp4j. It turns out that there were
many minor differences between the two modules, both typos and
mismatching types.

The new tests caught several issues that are fixed in this commit:

- remove `providesFileWatching`
- fix lsp4j `BuildServer.buildTargetInverseSources`
- add `BuildTarget.dependencies` to the spec
- make `BuildTarget.languageIds` required
- fix lsp4j `InverseSourcesParams.textDocument` type
- make all `BuildServerCapability` fields optional
- make lsp4s `BuildTarget.displayName` optional
- fix lsp4s `ResourcesResult.uris`
- add missing lsp4s endpoints `BuildTarget.{resources,didChange}`
- fix lsp4s endpoint `BuildTarget.sources`
- add constructors for all fields in lsp4j, more convenient than
  using setters.

This PR may have one controversial change: `{CompileParams,RunParams,TestParams}.arguments` are now typed as `Option[List[String]]`. I'm open for suggestions, it was inconsistent before in spec + bsp4s + bsp4j.

Fixes #57

### CODE REVIEW

**Comment by olafurpg** [comment]
Found a way in  2f229c2 to simplify the tests while making them more robust and test a larger surface 🎉

File: bsp4s/src/main/scala/ch/epfl/scala/bsp/Bsp.scala [hunk 1]
```diff
@@ -303,20 +302,20 @@ case object BuildTargetEventKind {
...
+    target: BuildTargetIdentifier,
+    resources: List[Uri]
 )
 
 // Request: 'buildTarget/compile', C -> S
 @JsonCodec final case class CompileParams(
     targets: List[BuildTargetIdentifier],
     originId: Option[String],
-    arguments: List[Json]
+    arguments: Option[List[String]]
```
Comment (jastice) [inline bsp4s/src/main/scala/ch/epfl/scala/bsp/Bsp.scala]: in Java version this is `@NonNull`, so why `Option` here?

File: bsp4s/src/main/scala/ch/epfl/scala/bsp/Bsp.scala [hunk 2]
```diff
@@ -336,7 +335,7 @@ case object BuildTargetEventKind {
 @JsonCodec final case class TestParams(
     targets: List[BuildTargetIdentifier],
     originId: Option[String],
-    arguments: List[Json]
+    arguments: Option[List[String]]
```
Comment (jastice) [inline bsp4s/src/main/scala/ch/epfl/scala/bsp/Bsp.scala]: as above, `@NonNull` in Java

File: bsp4s/src/main/scala/ch/epfl/scala/bsp/Bsp.scala [hunk 3]
```diff
@@ -360,7 +359,7 @@ case object BuildTargetEventKind {
 @JsonCodec final case class RunParams(
     target: BuildTargetIdentifier,
     originId: Option[String],
-    arguments: List[Json]
+    arguments: Option[List[String]]
```
Comment (jastice) [inline bsp4s/src/main/scala/ch/epfl/scala/bsp/Bsp.scala]: as above, `@NonNull` in Java

File: docs/bsp.md [hunk 4]
```diff
@@ -326,37 +325,44 @@ trait InitializeBuildResult {
...
   /** The languages the server supports run via method buildTarget/run */
-  def runProvider: {
-    def languageIds: List[String]
-  }
+  def runProvider: Option[RunProvider]
   
   /** The server can provide a list of targets that contain a
     * single text document via the method buildTarget/inverseSources */
-  def inverseSourcesProvider: Boolean
+  def inverseSourcesProvider: Option[Boolean]
```
Comment (jastice) [inline docs/bsp.md]: is `@NonNull` in Java

File: docs/bsp.md [hunk 5]
```diff
@@ -326,37 +325,44 @@ trait InitializeBuildResult {
...
   
   /** The server provides sources for library dependencies
     * via method buildTarget/dependencySources */
-  def dependencySourcesProvider: Boolean
+  def dependencySourcesProvider: Option[Boolean]
```
Comment (jastice) [inline docs/bsp.md]: is `@NonNull` in Java

File: docs/bsp.md [hunk 6]
```diff
@@ -326,37 +325,44 @@ trait InitializeBuildResult {
...
   
   /** The server provides all the resource dependencies
     * via method buildTarget/resources */
-  def resourcesProvider: Boolean
+  def resourcesProvider: Option[Boolean]
```
Comment (jastice) [inline docs/bsp.md]: is `@NonNull` in Java

File: docs/bsp.md [hunk 7]
```diff
@@ -326,37 +325,44 @@ trait InitializeBuildResult {
...
   
   /** The server sends notifications to the client on build
     * target change events via buildTarget/didChange */
-  def buildTargetChangedProvider: Boolean
+  def buildTargetChangedProvider: Option[Boolean]
```
Comment (jastice) [inline docs/bsp.md]: is `@NonNull` in Java

File: docs/bsp.md [hunk 8]
```diff
@@ -928,8 +934,8 @@ trait RunParams {
     * The server may include this id in triggered notifications or responses. */
   def originId: Option[String]
   
-  /** Optional arguments to the test execution. */
-  def arguments: Option[Json]
+  /** Optional arguments to the executed application. */
+  def arguments: Option[List[String]]
```
Comment (jastice) [inline docs/bsp.md]: is `@NonNull` in Java

**Comment by jastice** [comment]
Okay then my question is why are all those fields optional now? :)

**Comment by olafurpg** [comment]
The capability provider fields became optional were made optional to make it possible to remove them in future releases without breaking older clients. This matches LSP capability fields https://microsoft.github.io/language-server-protocol/specification#initialize

**Comment by jastice** [comment]
I'm not a huge fan of the approach but let's go with established practices here 👍

**Pull request merged**